[IMP] purchase, account, sale: add account.order(.line).mixin

R1D1CUL0US opened the pull request:
A lot of behavior is shared between the sale order (line) and purchase order (line). To avoid having to maintain this code in parallel and allow for sharing of the overlapping code, we introduce two new mixins:
1. account.order.mixin, for sale.order and purchase.order
2. account.order.line.mixin, for sale.order.line and purchase.order.line

We make use of this opportunity to also rename some of the shared fields to conform to the naming standards.

Additionally, this PR adds the option to create down payments for purchase orders by moving the existing functionality from sale orders into the newly created mixins. This new function is disabled by default for purchase, but can be turned on with a new checkbox in the settings.

task-3600910

Enterprise PR: https://github.com/odoo/enterprise/pull/55711
Upgrade PR: https://github.com/odoo/upgrade/pull/5627

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/144608).

Feyensv reviewed (commented):
Not really convinced by the `account.downpayment` and `account.downpayment.parent` names for the mixins, can't we go with something more global, maybe something like `account.invoiceable.order` and `account.invoiceable.line` , and lay the groundwork for a future harmonization of the logic shared by SO & PO's (fpos, taxes, payment terms, currency and rate, ...)

R1D1CUL0US commented:
> Not really convinced by the `account.downpayment` and `account.downpayment.parent` names for the mixins, can't we go with something more global, maybe something like `account.invoiceable.order` and `account.invoiceable.line` , and lay the groundwork for a future harmonization of the logic shared by SO & PO's (fpos, taxes, payment terms, currency and rate, ...)

@Feyensv Thank you for your review! I have implemented your suggested changes and renamed the mixins as suggested.

Feyensv reviewed (commented):
Pushed a little fixup commit to mostly reorder some methods ;)
Thanks for the changes, I feel like this will become a really nice base for future improvements :).

R1D1CUL0US commented on addons/purchase/wizard/purchase_make_bill_advance_views.xml:
```diff
@@ -0,0 +1,22 @@
+        <field name="mode">primary</field>
+        <field name="name">Create Bills</field>
+        <field name="arch" type="xml">
+            <xpath expr="//form" position="attributes">
```
It is indeed easier to read, I replaced them where appropriate

R1D1CUL0US commented on addons/sale/wizard/sale_make_invoice_advance_views.xml:
```diff
@@ -1,73 +1,36 @@
         <field name="binding_model_id" ref="sale.model_sale_order"/>
         <field name="binding_view_types">list</field>
     </record>
-
```
I was a little too eager while deleting ;)

RicGR98 commented:
Hey :)
I've just launched the Entreprise build and it fails.
Could you check if this is linked to your PR ?
Thanks

h4818 reviewed (commented):
It seems I started a review and didn't finish/submit it. most is probably outdated, but I'll just submit anyway

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+
+    def _create_invoices(self, orders):
+        if self.advance_payment_method == 'delivered':
+            return orders._create_invoices(final=True, grouped=not self._invoice_grouped())
```
Double check that we really want this `grouped=not self._invoice_grouped()`

R1D1CUL0US replied:
It looks weird indeed (but is correct). I changed the name to consolidated billing to make it less confusing.

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+        values, accounts = self._prepare_down_payment_lines_values(order)
+        down_payment_lines = OrderLine.create(values)
+
+        invoice = self.env['account.move'].sudo().create(
+            self._prepare_invoice_values(order, down_payment_lines, accounts)
+        ).with_user(self.env.uid)  # Unsudo the invoice after creation
```
Why do we need sudo in the first place ?

R1D1CUL0US replied:
Added a reference

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+        if self.advance_payment_method == 'fixed':
+            delta_amount = (invoice.amount_total - self.fixed_amount) * (-1 if invoice.is_inbound() else 1) * self.order_ids._sign()
+            if not order.currency_id.is_zero(delta_amount):
+                receivable_line = invoice.line_ids \
+                                      .filtered(lambda aml: aml.account_id.account_type == self._account_type())[:1]
+                product_lines = invoice.line_ids \
+                    .filtered(lambda aml: aml.display_type == 'product')
+                tax_lines = invoice.line_ids \
+                    .filtered(lambda aml: aml.tax_line_id.amount_type not in (False, 'fixed'))
```
Peculiar formatting, I guess IDE reformatted the pasting ?

R1D1CUL0US replied:
Nope, came straight from the Sales code ;) But I put it back on one line.

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+                    })]
+                    delta_sign = 1 if delta_amount > 0 else -1
+                    for lines, attr, sign in (
+                            (product_lines, 'price_total', -1),
+                            (tax_lines, 'amount_currency', 1),
```
Too much indent

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+            )
+            for line in order_lines
+        ]
+        computed_taxes = self.env['account.tax']._compute_taxes(
+            tax_base_line_dicts)
```
one line

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+                # of the tax amount. Therefore fixed taxes are removed and are replace by a new line
+                # with appropriate amount, and non fixed taxes if the fixed tax affected the base of
+                # any other non fixed tax.
+                if fixed_tax.include_base_amount:
+                    pct_tax = taxes[list(taxes).index(fixed_tax) + 1:]\
+                        .filtered(lambda t: t.is_base_affected and t.amount_type != 'fixed')
+                else:
+                    pct_tax = self.env['account.tax']
+                down_payment_values.append([
+                    pct_tax,
+                    line['analytic_distribution'],
+                    line['quantity'] * fixed_tax.amount,
```
Feels a bit dangerous recalculating the taxes by hand, there is a method for that in `account.tax`: `compute_all`.

Maybe we can get @smetl 's opinion on this as he's working on a refactoring of the taxes

RicGR98 commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,193 @@
+
+    invoice_status = fields.Selection([])
+
+    def _qty_field_name(self):
+        raise NotImplementedError # To override
```
It's a bit sad that we don't have the same names across different modules and need this kind of method :/
[ignore this]

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+            'is_downpayment': True,
+            'sequence': order.order_line and order.order_line[-1].sequence + 1 or 10,
+        }
+        del context
```
?

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+            name = _("Down payment of %s%%", self.amount)
+        else:
+            name = _("Down Payment")
+        del context
```
I don't understaaaand why is this everywhereeeee and what it does. Should the context variable be assigned to the `self.env.context` ? And then reset back (not deleted) ?

Feyensv replied:
The translation utils takes the lang from that context dict.  It's magically ugly, but it's been like that for ages.
 https://github.com/odoo/odoo/blob/42b4944a586997c189a2bfc245600f73c48841d7/odoo/tools/translate.py#L472-L477

RicGR98 replied:
Thanks for the explanation ! :exploding_head:

RicGR98 commented on addons/sale/models/sale_order.py:
```diff
@@ -1420,6 +1350,21 @@ def _create_invoices(self, grouped=False, final=False, date=None):
             )
         return moves
 
+    def _filter_subrecords(self):
```
I'd add either a comment section so we understand where there methods come from
`##### Override methods of account.invoiceable.order ####`
Same for SaleOrderLine, PurchaseOrder(Line)

Feyensv replied:
It's in an invoicing section, which mostly contains overrides of the mixin, it's enough IMHO.

RicGR98 commented on addons/purchase/models/purchase_order.py:
```diff
@@ -555,39 +555,35 @@ def _add_supplier_to_product(self):
+            'account_id': False,
+            **optional_values
+        }
+        del context
```
Same as above

RicGR98 commented on addons/purchase/models/purchase_order_line.py:
```diff
@@ -31,6 +31,7 @@ class PurchaseOrderLine(models.Model):
         digits='Discount',
         store=True, readonly=False)
     taxes_id = fields.Many2many('account.tax', string='Taxes', domain=['|', ('active', '=', False), ('active', '=', True)], context={'active_test': False})
+    tax_id = fields.Many2many(related='taxes_id', string='Taxes (alias)')
```
```suggestion
    tax_ids = fields.Many2many(related='taxes_id', string='Taxes (alias)')
```

Feyensv replied:
Cosmetic changes are not in the scope of this PR (and will explode the size of the diff).

RicGR98 commented on addons/sale_timesheet/wizard/sale_make_invoice_advance.py:
```diff
@@ -46,6 +46,6 @@ def _create_invoices(self, sale_orders):
                 timesheet_start_date=self.date_start_invoice_timesheet,
                 timesheet_end_date=self.date_end_invoice_timesheet
-            )._create_invoices(final=self.deduct_down_payments)
+            )._create_invoices(final=True)
```
Is this intended ?

R1D1CUL0US replied:
The field was always True (appears to be leftover from some time ago when it was used) ... so I removed it

RicGR98 commented on addons/purchase/tests/test_purchase_downpayment.py:
Didn't check :see_no_evil: , I guess it's the same as `test_sale_downpayment`

R1D1CUL0US replied:
It indeed takes all of the tests from there + some more that were scattered throughout sales tests but also tested down payment functionality. I did simplify them a little by refactoring out common repeated parts into methods to make it cleaner but the tested functionality stays the same :)

RicGR98 commented on addons/purchase/models/purchase_order.py:
```diff
@@ -1070,3 +1068,32 @@ def _is_readonly(self):
+    def action_create_bill(self):
+        if self.user_has_groups('purchase.group_down_payment'):
+            return {
+                'name': _('Create invoices'),
```
Create bills or invoices ?

RicGR98 commented on addons/purchase/wizard/purchase_make_bill_advance.py:
```diff
@@ -0,0 +1,36 @@
+    amount_to_invoice = fields.Monetary(help="The amount to invoice = Purchase Order Total - Confirmed Down Payments.")
+    order_ids = fields.Many2many('purchase.order')
+
+    def view_draft_invoices(self):
```
invoices or bills ?
There are some other examples in the files, could you please double check all utilisations of invoices vs bills (and use account_move if you need to include both, for instance in the abstract mixin)

RicGR98 replied:
:up:

RicGR98 replied:
At least the ones that are displayed to the user, for instance ` _('Draft Invoices'),`

RicGR98 reviewed (commented):
Thanks for the PR ! :sun_with_face: 

Runbot is still red on the l10n builds

Feyensv commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,80 @@
+    order_id = fields.Many2one('account.invoiceable.order')
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
+    tax_id = fields.Many2many('account.tax')
```
You don't want to touch that right now.

R1D1CUL0US replied:
I'll leave it for now ;)

Feyensv commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,193 @@
+        pending_section = None
+        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
+
+        for line in self.order_line:
```
You won't change that either.

R1D1CUL0US commented on addons/stock_landed_costs/models/purchase.py:
```diff
@@ -3,7 +3,7 @@
 
-    def _prepare_account_move_line(self, move=False):
-        res = super()._prepare_account_move_line(move)
+    def _prepare_invoice_line(self, move=False, **optional_values):
```
So I renamed it to match the mixin method, which in turn matches the method in sales. So best to change all of them to _prepare_account_move_line then?

R1D1CUL0US commented on addons/account/models/account_downpayments.py:
```diff
@@ -0,0 +1,267 @@
+
+    # To override
+    invoice_lines = fields.One2many('account.move.line')
+    display_type = fields.Selection([
```
I moved it in the mixin as it depends on these fields being there to function, to make it clear it cannot just be renamed/changed for example in a class implementing the mixin

R1D1CUL0US commented on addons/account/models/account_downpayments.py:
```diff
@@ -0,0 +1,267 @@
+        """ Check whether given lines can be deleted or not.
+
+        * Down payment lines who have not yet been invoiced can always be deleted.
+        * Sections and Notes can always be deleted.
```
This was moved out of the mixin in the meanwhile

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,193 @@
+            # If the invoice status is 'Fully Invoiced' force the amount to invoice to equal zero and return early.
+            if order.invoice_status == 'invoiced':
+                order.amount_to_invoice = 0.0
+                return
```
I took this from the original sale order code but indeed it looks incorrect, I'll change it

R1D1CUL0US commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,80 @@
+            if not line.product_id:
+                continue
+            lang = line.order_id._get_lang()
+            if lang != self.env.lang:
+                line = line.with_context(lang=lang)
```
Yes but we need the lang again on line 63

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,193 @@
+from odoo.tools import float_is_zero
+
+
+class InvoiceableOrder(models.AbstractModel):
+    _name = 'account.invoiceable.order'
```
From my experience in Logistics (purchase scope) there are a lot of cases where we have overlapping functionality and are maintaining/fixing a lot of similar stuff between sales and purchase. As an example, my original PR was a full copy of all of the down payment stuff from sales. That's why I personally think the purpose of these classes is to build a good foundation for adding more overlapping sale/purchase code in the future and to maintain it in a single place instead of in parallel. So I think the names InvoiceableOrder and InvoiceableLine are quite good.

RicGR98 replied:
Agreed after reading  https://github.com/odoo/odoo/pull/144608#pullrequestreview-1829400123 but why not keeping the mixin keyword ? It'd be much easier to understand that this is not an object on its own.

R1D1CUL0US replied:
![image](https://github.com/odoo/odoo/assets/3251665/c66446ec-85d6-4ef6-a144-79b353f247aa)

R1D1CUL0US commented on addons/account/models/account_move.py:
```diff
@@ -4276,17 +4281,29 @@ def action_post(self):
+        # Update any linked downpayment lines
+        downpayment_lines = self.line_ids._get_linked_downpayment_lines().filtered(lambda l: l.is_downpayment and not l.display_type and not l.order_id._is_locked())
+        other_order_lines = downpayment_lines.order_id.order_line - downpayment_lines
+        real_invoices = set(other_order_lines.invoice_lines.move_id)
```
https://github.com/odoo/odoo/pull/145964#discussion_r1446249039

R1D1CUL0US commented on addons/sale/wizard/sale_make_invoice_advance.py:
```diff
@@ -10,339 +10,43 @@
-    # Down Payment logic
-    has_down_payments = fields.Boolean(
-        string="Has down payments", compute="_compute_has_down_payments")
-    deduct_down_payments = fields.Boolean(string="Deduct down payments", default=True)
```
The deduct_down_payments field appears to be left over in the code but has been removed from the view and is always True, so I removed it.

R1D1CUL0US commented on addons/account/models/account_move.py:
```diff
@@ -2756,9 +2756,14 @@ def _unlink_forbid_parts_of_chain(self):
 
     def unlink(self):
+        downpayment_lines = (self.mapped('line_ids')._get_linked_downpayment_lines()
+                             .filtered(lambda line: line.is_downpayment and line.invoice_lines <= self.mapped('line_ids')))
```
I improved the indentation as per your suggested change and added a reference to the original PR that introduced the last filter condition

R1D1CUL0US commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,338 @@
+
+
+class AdvancePayment(models.AbstractModel):
+    _name = 'account.advance.payment'
```
I'll change it for purchase and the mixin but I'm very hesitant to do it for sales... It's referenced everywhere, the diff will be huge...

RicGR98 reviewed (commented):
Could you check these last 2 comments ? Thanks

\+ Runbot l10n is red, don't forget to run the l10n and enterprise builds :)

oco-odoo commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,79 @@
+    tax_id = fields.Many2many('account.tax')
+
+    def _get_invoice_lines(self):
+        raise NotImplementedError # To override
```
sale and purchase seem to override it with the exact same function ; why no implementing it here directly ?

oco-odoo commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,79 @@
+        raise NotImplementedError # To override
+
+    def _get_description(self):
+        raise NotImplementedError # To override
```
All those hook functions need a proper docstring to explain when they are called and to which purpose.

oco-odoo commented on addons/purchase/models/purchase_order_line.py:
```diff
@@ -558,9 +576,12 @@ def _prepare_account_move_line(self, move=False):
             'price_unit': self.currency_id._convert(self.price_unit, aml_currency, self.company_id, date, round=False),
             'tax_ids': [(6, 0, self.taxes_id.ids)],
             'purchase_line_id': self.id,
```
A lot of the things set here seem to be common between sale and purchase ; why not setting them in the mixin, and only add the required stuff in each of so/po lines ?

oco-odoo commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,192 @@
+
+
+class InvoiceableOrderMixin(models.AbstractModel):
+    _name = 'account.invoiceable.order'
```
Wouldn't 'account.order' be explicit enough ?  :thinking: Same for order lines

oco-odoo commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,192 @@
+    def _create_invoices(self, grouped=False, final=False, date=None):
+        raise NotImplementedError # To override
+
+    def action_view_invoice(self, invoices=False):
```
Here as well, those functions need docstrings

oco-odoo replied:
action_view_invoice has to choose a different tree view to open between sale and purchase, so it makes sense to have a hook to override for this. Though, it also sets default values in the context in sale than in purchase ; shouldn't these behaviors be aligned (putting nearly everything into the mixin) ? I don't see why we need those default values in sale tbh :thinking:

oco-odoo commented on addons/purchase/models/purchase_order.py:
```diff
@@ -1069,3 +1070,32 @@ def _is_readonly(self):
+        return 'product_qty'
+
+    def _sign(self):
+        return +1
```
```suggestion
        return 1
```

oco-odoo commented on addons/account/models/account_move_line.py:
```diff
@@ -415,6 +415,7 @@ class AccountMoveLine(models.Model):
              "associated partner",
     )
     is_refund = fields.Boolean(compute='_compute_is_refund')
+    is_downpayment = fields.Boolean()
```
Seems weird to add a stored field on all the move lines just for that. Is there no other way ? Can't it be computed ? What kind of processing do you apply with that field after the invoice has been created ? :thinking:

R1D1CUL0US replied:
We already are storing this field on all account move lines like this as soon as you install the sale module

R1D1CUL0US replied:
I'm not really sure how we would be able to guess it from the account and balance sign as it's just on the Product Sales account? The display_type is currently set to product, I could add a display_type of down_payment but I'm not sure if this is a better solution than the current boolean as we would need to change a lot of the checks of type 'display_type == product' to also include lines with display type of down_payment. (but maybe I'm missing something?)

oco-odoo replied:
The move lines table is huge (a medium-sized accounting contains millions of records in that one) ; that's why I'd prefer to avoid adding one field here :thinking: But for now, I suppose it's not much worse than it used to be (since this field was in sale anyway). So let's keep it this way. But it keep in mind if would be nice to remove it in the future if we can.

oco-odoo commented on addons/account/models/account_move.py:
```diff
@@ -4299,17 +4311,38 @@ def action_post(self):
             other_moves._post(soft=False)
-        return False
+        res = False
+        # Update any linked downpayment lines
```
Wait ... you update the SO from the invoice when posting it, is that it ? Isn't too much magic killing the magic ?

R1D1CUL0US replied:
Yes this is how it works right now for SOs and I generalized it to also apply to POs in the same way

oco-odoo commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,335 @@
+    def _account_category(self):
+        raise NotImplementedError # To override
+
+    def _taxes_field_name(self):
```
Same here: give those functions docstrings. Also, your other mixins tend to start with the functions to override ; would be clearer if you did the same here

oco-odoo reviewed (changes requested):
I had a very quick look at the accounting part ; here are a few points of attention. I don't have much time ATM, so I directly asked when something seemed fishy instead of searching in depth

R1D1CUL0US commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,80 @@
+                                         " They are not copied when duplicating an order.")
+
+    # To override
+    invoice_lines = fields.One2many('account.move.line')
```
Changing this would have very far-reaching consequences throughout multiple modules across accounting, sales and logistics team scopes. So my counter-proposal would maybe be to do this renaming in a separate refactor PR instead of this one?

oco-odoo commented on addons/purchase/models/purchase_order.py:
```diff
@@ -1076,3 +1077,32 @@ def _is_readonly(self):
+        return self.action_create_invoice(True)
+
+    def _qty_field_name(self):
+        return 'product_qty'
```
product_qty and product_uom_qty should be merged and declared on the parent class.

R1D1CUL0US commented on addons/account/models/account_invoiceable_line.py:
I had it as a single file first, but the reviewer from sales requested it to be split into two files

R1D1CUL0US commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,117 @@
+            return _("Down Payments")
+
+        dp_state = self._get_downpayment_state()
+        name = _lt("Down Payment")
```
It was introduced here: https://github.com/odoo/odoo/commit/5f27513bf7f955419059bdd95e23a1f829696f98#diff-9d184722ff87da4215bc50fa2e539c0f6c2f099c831aa398e593220cf1325640. I think it is done as we just use it as a default value here and we might not need the translation when overwriting the name later in the code

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+            'account_id': False,
+            **optional_values
+        }
+        del context
```
https://github.com/odoo/odoo/commit/964d35207717af9e7bf42e6f9293249cb6e48991 I'll add in a reference to this commit

R1D1CUL0US commented on addons/purchase/models/purchase_order.py:
```diff
@@ -152,6 +152,9 @@ def _compute_invoice(self):
     receipt_reminder_email = fields.Boolean('Receipt Reminder Email', compute='_compute_receipt_reminder_email')
     reminder_date_before_receipt = fields.Integer('Days Before Receipt', compute='_compute_receipt_reminder_email')
 
+    amount_to_invoice = fields.Monetary("Amount to be billed")
+    amount_invoiced = fields.Monetary("Amount already billed")
```
It is generalized in account.order, here it's just updating the string to say billed instead of invoiced. Would you prefer if the strings would be the same on the fields in purchase and sales?

oco-odoo replied:
If it's functionally needed, alrigth :) Maybe just put a small comment on the same line telling that on PO/SO, so that it's more explicit for a reader not paying enough attention as myself :see_no_evil:

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+    invoice_ids = fields.Many2many('account.move')
+    order_line = fields.One2many('account.order.line')
+
+    invoice_status = fields.Selection([])
```
Yes but they have different strings, so if I declare them here they will need to share the same user-facing name, no?

R1D1CUL0US commented on addons/sale/models/sale_order_line.py:
```diff
@@ -1170,33 +1127,21 @@ def _set_analytic_distribution(self, inv_line_vals, **optional_values):
-        """
-        return {}
-
     def _validate_analytic_distribution(self):
```
For Sale Order Lines we are filtering out all lines that are not in a draft or sent state, but on the purchase side we are always running the validation, so I can merge them but will need to create a new method then to override and filter the records I think. Like _should_validate_analytic_distribution() maybe?

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+    amount_total = fields.Monetary()
+
+    invoice_ids = fields.Many2many('account.move')
+    order_line = fields.One2many('account.order.line')
```
Yes it was some kind of "abstract declaration" because it felt weird to use an undeclared field in the abstract class methods, like magically assuming that all subclasses will just have a field with that name pointing to some subclass of the abstract class. But as it is useless/not necessary would you prefer me to remove it here and just keep it on the subclasses?

oco-odoo replied:
Hmmmm :thinking: Not sure. I don't know if we had other cases like that in the past. I suggest you ask the ORM team what they think about it. Maybe there's just a function we can override to check that the proper fields are declared on the subclasses ? (I doubt it)

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+
+            order.amount_to_invoice = order.amount_total
+            for invoice in order.invoice_ids.filtered(lambda x: x.state == 'posted'):
+                prices = sum(invoice.line_ids.filtered(order._filter_subrecords()).mapped('price_total'))
```
It does look a little strange indeed, the original reason for the filter was added here in sales: https://github.com/odoo/odoo/commit/e2ec9e6c86949f5781d03a7dbb1065ef8af3196a

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+                ]
+        return vals_list
+
+    def _generate_invoice_values(self, final=False):
```
It is being called from the purchase_order and sale_order implementations of _create_invoices

R1D1CUL0US commented on addons/account/models/account_invoiceable_order.py:
```diff
@@ -0,0 +1,226 @@
+            for invoice in order.invoice_ids.filtered(lambda x: x.state == 'posted'):
+                prices = sum(invoice.line_ids.filtered(order._filter_subrecords()).mapped('price_total'))
+                invoice_amount_currency = invoice.currency_id._convert(
+                    prices * invoice.direction_sign * order._sign(),
```
I would be hesitant to change this as there is indeed the complexity with credit notes, taxes and rounding to take into account. There was already a bugfix because it was not working properly in the credit note case...

oco-odoo replied:
Alright, let's forget that for now. Just keep it in mind for the future, maybe :thinking:

RicGR98 commented on addons/account/models/account_order.py:
```diff
@@ -0,0 +1,229 @@
+    amount_total = fields.Monetary()
+
+    invoice_ids = fields.Many2many('account.move')
+    order_line = fields.One2many('account.order.line')
```
```suggestion
    order_line_id = fields.One2many('account.order.line')
```

R1D1CUL0US replied:
![image](https://github.com/odoo/odoo/assets/3251665/3c87995e-43d8-4ef0-b331-b303a26c0e38)

RicGR98 replied:
I doubt all these are correct matches of the field, and you already did rename `invoice_lines` to `invoice_line_ids` so why not... I don't really care but I'll let the final reviewer decide

RicGR98 commented on addons/account/models/account_order.py:
```diff
@@ -0,0 +1,229 @@
+
+    amount_total = fields.Monetary()
+
+    invoice_ids = fields.Many2many('account.move')
```
I thought oco suggested to use `account_move` or `move` instead of invoice ? (this commit is actually valid for all other usages of invoices)

RicGR98 commented on addons/account/models/account_order_line.py:
```diff
@@ -0,0 +1,109 @@
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
+    tax_ids = fields.Many2many('account.tax')
+    product_uom = fields.Many2one('uom.uom')
```
```suggestion
    product_uom_id = fields.Many2one('uom.uom')
```

RicGR98 commented on addons/account/models/account_order_line.py:
```diff
@@ -0,0 +1,109 @@
+
+    def _compute_name(self):
+        for line in self:
+            if line.is_downpayment:
```
no else even though it's a non-stored computed field ?

R1D1CUL0US replied:
In the else case it is set in an override in the subclass.

RicGR98 commented on addons/account/models/account_order_line.py:
```diff
@@ -0,0 +1,109 @@
+        for line in self:
+            if line.is_downpayment:
+                lang = line.order_id._get_lang()
+                if lang != self.env.lang:
+                    line = line.with_context(lang=lang)
```
Isn't this similar to this ?
```suggestion
                line = line.with_context(lang=lang)
```

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,339 @@
+                        (tax_lines, 'amount_currency', 1),
+                    ):
+                        remaining = delta_amount
+                        lines_len = len(lines)
```
i'd inline it

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,339 @@
+        )
+
+        # Ensure the invoice total is exactly the expected fixed amount.
+        if self.advance_payment_method == 'fixed':
```
No need for percentage ?

R1D1CUL0US replied:
In the percentage case this issue with the rounding and the tax does not occur. It is only to make sure that in the fixed case it does not end up being 250.01 when putting 250.

RicGR98 commented on addons/account/wizard/accrued_orders.py:
```diff
@@ -179,13 +179,14 @@ def _ellipsis(string, size):
                         amount_currency = order_line.untaxed_amount_to_invoice
                         amount = order.currency_id._convert(amount_currency, self.company_id.currency_id, self.company_id)
-                        fnames = ['qty_to_invoice', 'untaxed_amount_to_invoice', 'qty_invoiced', 'qty_delivered', 'invoice_lines']
+                        fnames = ['qty_to_invoice', 'untaxed_amount_to_invoice', 'qty_invoiced', 'qty_delivered',
```
one line

RicGR98 commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,339 @@
+        if self.advance_payment_method == 'percentage':
+            ratio = self.amount / 100
+        else:
+            ratio = self.fixed_amount / order.amount_total if order.amount_total else 1
```
Should we add a warning if the fixed amount is bigger than the total amount ?

R1D1CUL0US replied:
There already is :)
![image](https://github.com/odoo/odoo/assets/3251665/9ef2a21f-73f0-43a5-9d60-8ab2435ac474)

R1D1CUL0US commented on addons/account/models/account_order.py:
```diff
@@ -0,0 +1,229 @@
+        )
+
+    def copy_data(self, default=None):
+        default = dict(default or {})
```
It is different when default is not a dict but for example False. I see this pattern everywhere in the overrides of copy_data

RicGR98 replied:
```sh
>>> dict(False or {})
{}
>>> dict(None or {})
{}
>>> dict({"rigr": "odoo"} or {})
{'rigr': 'odoo'}
>>> 
>>> False or {}
{}
>>> None or {}
{}
>>> {"rigr": "odoo"} or {}
{'rigr': 'odoo'}
```

R1D1CUL0US commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,339 @@
+
+    # === CONSTRAINT METHODS ===#
+
+    def _check_amount_is_positive(self):
+        for wizard in self:
+            if (wizard.advance_payment_method == 'percentage' and float_compare(wizard.amount, 0, precision_rounding=wizard.currency_id.rounding) <= 0) \
+                    or (wizard.advance_payment_method == 'fixed' and float_compare(wizard.fixed_amount, 0, precision_rounding=wizard.currency_id.rounding) <= 0):
+                raise UserError(_('The value of the down payment amount must be positive.'))
```
https://github.com/odoo/odoo/pull/112905#discussion_r1130670664

oco-odoo commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,117 @@
+    invoice_lines = fields.One2many('account.move.line')
+    display_type = fields.Selection([
+        ('line_section', "Section"),
+        ('line_note', "Note")], default=False, help="Technical field for UX purpose.")
```
I had a look :monocle_face: 

For POL, it doesn't work because of this line https://github.com/odoo/odoo/blob/17.0/addons/purchase/models/purchase_order_line.py#L197 . If you remove the explicit default in the field declaration, the dict returned by default_get will not contain 'display_type', and you'll get an error. Though, this line seems an absolute nonsense, since the default_get will always return False, so essentially, we'll only enter the condition if values.get('display_type') is evaluated to True. 

For SOL, I noticed your branch was declaring the field a second time with the same values as in account.order.line.mixin ; I guess it's a leftover, you should remove it. But apart from that, I don't have any error when creating them :thinking:

oco-odoo commented on addons/account/models/account_invoiceable_line.py:
```diff
@@ -0,0 +1,117 @@
+        ('line_note', "Note")], default=False, help="Technical field for UX purpose.")
+    product_id = fields.Many2one('product.product')
+    name = fields.Text(compute='_compute_name')
+    order_id = fields.Many2one('account.order')
```
It's still there ; the comment is marked outdated because the model was renamed to account.order.mixin. Did you forget, or is it intended ? If so, why ?

oco-odoo commented on addons/account/models/account_order_mixin.py:
```diff
@@ -0,0 +1,230 @@
+    invoice_ids = fields.Many2many('account.move')
+    order_line = fields.One2many('account.order.line.mixin')
+
+    invoice_status = fields.Selection([])
```
The common values for the selection could probably be defined here, just for clarity. You'll change some labels in the override anyway

R1D1CUL0US replied:
Yeah they have some common values indeed but they have different labels (ex. 'nothing to bill' vs 'nothing to invoice'), so I'm not sure it is possible to define them on the mixin?

oco-odoo replied:
You can define the common things on the mixin, and only the label in each of the modules :)

oco-odoo commented on addons/account/models/account_order_mixin.py:
```diff
@@ -0,0 +1,230 @@
+    order_line = fields.One2many('account.order.line.mixin')
+
+    invoice_status = fields.Selection([])
+    name = fields.Char()
```
SO and PO declarations seem the same for this one as well

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    _name = 'account.order.line.mixin'
+    _description = 'Account Order Line Mixin'
+
+    is_downpayment = fields.Boolean(string="Is a down payment",
+                                    help="Down payments are made when creating account moves from an order."
+                                         " They are not copied when duplicating an order.",
+                                    )
```
```suggestion
    is_downpayment = fields.Boolean(
        string="Is a down payment",
        help="Down payments are made when creating account moves from an order. They are not copied when duplicating an order.",
    )
```

Sale only overrides this field to say "sales order" in the helper ; we don't care, do we ?

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+        default=False,
+        help="Technical field for UX purpose.",
+    )
+    product_id = fields.Many2one('product.product')
```
Everything but the domain is common for this field, and should be declared here

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+                                    )
+
+    # To override
+    move_line_ids = fields.One2many('account.move.line')
```
Same as the relatrional fields to AbstractModels ; I'd like to have the ORM Team's opinion on that :thinking:

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+                                         " They are not copied when duplicating an order.",
+                                    )
+
+    # To override
```
Written like this, it gives the impression it targets the whole block, while it actually only concerns move_line_ids

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+        help="Technical field for UX purpose.",
+    )
+    product_id = fields.Many2one('product.product')
+    name = fields.Text(compute='_compute_name')
```
Fully identical declarations between sale and purchase; let's put it all here

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    product_id = fields.Many2one('product.product')
+    name = fields.Text(compute='_compute_name')
+    order_id = fields.Many2one('account.order.mixin')
+    qty_to_invoice = fields.Float()
```
The full field should be declared here, with nothing in sale/purchase. They can then both override its compute function (which can be empty here)

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    name = fields.Text(compute='_compute_name')
+    order_id = fields.Many2one('account.order.mixin')
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
```
Is really it intentional one of them is precompute and not the other ? Maybe a common declaration is possible, with an override of the compute ? :thinking:

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    order_id = fields.Many2one('account.order.mixin')
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
+    tax_ids = fields.Many2many('account.tax')
```
Purchase is missing check_company here as well ; should be in your branch fix for 16.4 as well. 

Everything that can be common should be done here, leaving only an override in sale for the compute part.

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
+    tax_ids = fields.Many2many('account.tax')
+    product_uom = fields.Many2one('uom.uom')
```
product_uom_id

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    qty_to_invoice = fields.Float()
+    price_unit = fields.Float()
+    tax_ids = fields.Many2many('account.tax')
+    product_uom = fields.Many2one('uom.uom')
```
Here as well, things should be factorized. No ondelete='restrict' in purchase ?

oco-odoo replied:
product_uom_id btw ^^

oco-odoo replied:
Also, shouldn't the ondelete='restrict' used in sales be kept in the generic mixin ? (it's not by default since the field is not required, I think ; its "set null"). If I try deleting a UOM that was used on some SO/PO, I expect the operation to fail, don't I ? ? :thinking:

oco-odoo commented on addons/account/models/account_order_line_mixin.py:
```diff
@@ -0,0 +1,111 @@
+    price_unit = fields.Float()
+    tax_ids = fields.Many2many('account.tax')
+    product_uom = fields.Many2one('uom.uom')
+    discount = fields.Float()
```
Maybe here as well, fully declaring here, and overriding the compute would be possible ?

Feyensv commented:
Really? Honestly, those last renames will break everything for nearly no gain ...
I can accept some fields being harmonized between PO & SO, POL & SOL to ease the development, but modifying the name of fields that are already identical between sale & purchase is going too far here IMHO.

R1D1CUL0US commented:
> Really? Honestly, those last renames will break everything for nearly no gain ... I can accept some fields being harmonized between PO & SO, POL & SOL to ease the development, but modifying the name of fields that are already identical between sale & purchase is going too far here IMHO.

I agree

oco-odoo commented:
> Really? Honestly, those last renames will break everything for nearly no gain ... I can accept some fields being harmonized between PO & SO, POL & SOL to ease the development, but modifying the name of fields that are already identical between sale & purchase is going too far here IMHO.

You're talking about product_uom and order_line, right ?

Why would it "Breaking everything" ? Seems a big overstatement.

To me, it's been like that forever, we've been procrastinating that cleaning since then, and it's just a good opportunity :shrug: I had a look at the code, and just a few searches with regexes can give all the occurrences we want pretty easily. It's not hard to do.

Feyensv commented:
> You're talking about product_uom and order_line, right ?

Yes

> Why would it "Breaking everything" ? Seems a big overstatement.

Any existing code relying on those fields has to be rewritten/adapted, including server actions, filters, ...
Part of this will be handled with upgrade scripts, but not all.
Also, any untested code will be forward-ported and break without any notice.
I'm sure internal team will love such a change too.

> To me, it's been like that forever, we've been procrastinating that cleaning since then, and it's just a good opportunity 🤷 I had a look at the code, and just a few searches with regexes can give all the occurrences we want pretty easily. It's not hard to do.

Considering we just discarded a dumb 'utf-8 removal' PR because it annoyed some people, when it didn't cost anything while being clearer, merging this kind of change where the gain is low while the cost isn't, isn't worth it IMO.
I agree with the renaming of very confusing fields (like the `product.supplierinfo.name` which was previously a `partner_id` IIRC) but in this case, for me, `order_line` and `product_uom` are not worth renaming.

oco-odoo commented:
> Considering we just discarded a dumb 'utf-8 removal' PR because it annoyed some people, when it didn't cost anything while being clearer, merging this kind of change where the gain is low while the cost isn't, isn't worth it IMO. I agree with the renaming of very confusing fields (like the `product.supplierinfo.name` which was previously a `partner_id` IIRC) but in this case, for me, `order_line` and `product_uom` are not worth renaming.

I still don't agree, but I see your point. Since your team is the one that'll get mostly impacted by this renaming, I won't fight for it. We can keep product_uom and order_line if you want. Sorry for the back and forth, @R1D1CUL0US ^^

oco-odoo reviewed (approved):
Nothing more to say on the accounting part ; to me we can move on with this :)

ticodoo reviewed (commented):
Just nitpicks so far that can be ignored (except the better description for the mixin). Will review some more tomorrow + you need to rebase due to conflicts :crying_cat_face:

ticodoo commented on addons/account/models/account_order_mixin.py:
```diff
@@ -0,0 +1,249 @@
+        """
+        raise NotImplementedError  # To override
+
+    def action_view_invoice(self, invoices=False):
```
Order the methods in this mixin according to the [Model attribute order](https://www.odoo.com/documentation/17.0/contributing/development/coding_guidelines.html#symbols-and-conventions) standard

ticodoo reviewed (commented):
I will continue the downpayment/refactoring adventure tomorrow :crossed_swords:

R1D1CUL0US commented on addons/sale/models/sale_order_line.py:
```diff
@@ -117,59 +101,34 @@ class SaleOrderLine(models.Model):
         compute='_compute_amount',
         store=True, precompute=True)
-    price_tax = fields.Float(
-        string="Total Tax",
+    price_total = fields.Monetary(
         compute='_compute_amount',
         store=True, precompute=True)
-    price_total = fields.Monetary(
-        string="Total",
+    price_tax = fields.Monetary(
         compute='_compute_amount',
         store=True, precompute=True)
```
I had this before but unfortunately this results in this error:
![image](https://github.com/odoo/odoo/assets/3251665/7903137d-24d4-4f6e-92b4-5dfe1a37ffa1)
So I added back in the other attributes

ticodoo reviewed (commented):
Almost done :skull: Will continue the good fight tomorrow.

A few of my comments are maybe for the sales team so feel free to ping them for those specific ones.

R1D1CUL0US commented on addons/account/wizard/account_advance_payment.py:
```diff
@@ -0,0 +1,344 @@
+
+    # === ONCHANGE METHODS ===#
+
+    @api.onchange('advance_payment_method')
+    def _onchange_advance_payment_method(self):
+        if self.advance_payment_method == 'percentage':
+            amount = self.default_get(['amount']).get('amount')
+            return {'value': {'amount': amount}}
```
It looks strange indeed... Apparently it is for default values set in the debug menu? (ref: https://github.com/odoo/odoo/pull/45460) But not sure if that is still a relevant use case

R1D1CUL0US commented on addons/account/models/account_order_mixin.py:
```diff
@@ -0,0 +1,249 @@
+        """
+        raise NotImplementedError  # To override
+
+    def _filter_subrecords(self):
+        """
+        Returns a filter function that only returns True for invoice lines linked to this order.
+        Used to compute the amount to invoice.
+        """
+        raise NotImplementedError  # To override
```
Yeah I had to re-split the function that used it between sales and purchase after some new commits in sales, so it's useless now indeed

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        }
+        downpayment = cls.env['purchase.advance.payment.wizard'].with_context(cls.context).create(payment_params)
+        downpayment.create_invoices()
+        cls.purchase_order.button_confirm()
```
It feels weird to me to directly reference the cls.record in a helper method like this. It's maybe better/safer to pass it as an argument every time to avoid unexpected behavior, e.g. someone calls the method and then tries to edit the record without realizing its changed due to the method. I'm not not sure if it may lead to unexpected behavior between rollbacks

ticodoo commented on addons/purchase/views/portal_templates.xml:
```diff
@@ -308,9 +309,9 @@
                                     </td>
                                     <td t-if="not update_dates and order.state in ['purchase', 'done']" t-attf-class="text-end {{ 'd-none d-md-table-cell' if report_type == 'html' else '' }}">
-                                        <span t-esc="', '.join(map(lambda x: (x.description or x.name), line.taxes_id))"/>
+                                        <span t-esc="', '.join(map(lambda x: (x.description or x.name), line.tax_ids))"/>
```
may as well clean it up while you're changing the line
```suggestion
                                        <span t-out="', '.join(map(lambda t: (t.description or t.name), line.tax_ids))"/>
```

ticodoo commented on addons/purchase/views/purchase_views.xml:
```diff
@@ -243,10 +244,10 @@
                                            groups="analytic.group_analytic_accounting"
                                            options="{'product_field': 'product_id', 'business_domain': 'purchase_order', 'amount_field': 'price_subtotal'}"/>
-                                    <field name="product_qty"/>
+                                    <field name="product_qty" readonly="is_downpayment"/>
```
You might need to add in extra `is_downpayment` logic for `date_planned`and `discount`. Should be double checked, I'm getting weird "Invalid Fields" warning if I try to set these values in my localhost + save

ticodoo reviewed (commented):
Still gotta look at the tests :skull: :skull:

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        result = self.env['purchase.order'].search(domain)
+        self.assertEqual(result, expected_result, "Unexpected result on search orders")
+
+    def test_downpayment(self):
```
Based on the tests that follows, seems like this should be named:

`test_downpayment_fixed_tax_incl`

Although am I missing something or don't all of these tests have tax included anyways because of how the class setup is designed?

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+
+        invoice = self.purchase_order.invoice_ids[0]
+        downpayment_aml = invoice.line_ids.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))[0]
+        self.assertEqual(downpayment_aml.price_total, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+        self.assertEqual(downpayment_aml.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+        self.post_bills()
+        self.assertEqual(downpayment_line.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
```
Make the errors more specific to whether you're looking at a PO or the Invoice line and what amount you're referring to. Also, I don't understand why you check the POL amount and not the Invoice lines values after posting the bills values. Does posting affect the PO? I also don't understand why you check posting the bill + the values afterwards for the percentage case but not the fixed amount down payment case

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        self.assertEqual(downpayment_line.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+
+    def test_downpayment_invoice_and_partial_credit_note(self):
+        """This test check that the downpayment line amount on the purchase order remains consistent"""
```
Test description doesn't match what the test is checking. Also it seems like the consistent dp line amount check was in the previous test?

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+
+        self.make_downpayment(advance_payment_method='fixed', fixed_amount=100)
+
+        # Ensure the downpayment line on the sale order is correctly set to 100
```
`sale order => purchase order`

I guess that it's supposed to be a refund in the comment + assert message below as well

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+    def test_multi_company_invoice(self):
+        """
+            Checks that the company of the invoices generated in a multi-company environment using the
+           'purchase.advance.payment.wizard' wizard fit with the company of the PO and not with the current company.
```
`'purchase.advance.payment.wizard' wizard matches the company of the PO and not with the current company.`

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+
+        # Make sure the invoice is not created with a journal in the context
+        # Because it makes the test always succeed (by using the journal company instead of the env company)
+        no_journal_ctxt = dict(self.context)
+        no_journal_ctxt.pop('default_journal_id', None)
+        no_journal_ctxt.pop('journal_id', None)
```
is this context supposed to be used somewhere?

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        pol_downpayment = self.purchase_order.order_line[2]
+        dp_invoice = self.purchase_order.invoice_ids[0]
+        dp_invoice.invoice_date = datetime.datetime.today()
+        dp_invoice.action_post()
```
there's only 1 invoice at this point, right? So `post_bills()`?

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+    @classmethod
+    def post_bills(cls):
+        cls.purchase_order.invoice_ids.write({'invoice_date': datetime.datetime.today()})
+        cls.purchase_order.invoice_ids.action_post()
```
Not sure if this method adds much, especially since I don't know what the behavior is if you call it after some invoices have already been posted. So it's not very re-usable depending on the test and only saves 1 line per call

Also, style nitpick:
`purchase_order.invoice_ids.invoice_date = datetime.datetime.today()`

ticodoo reviewed (commented):
The tests seem like they could be cleaned up a bit for purchase. If it's possible to combine some of the tests, that would be ideal for readability/cleanliness. I didn't finish going through all of them, but it seems like some are inconsistent with what they do or don't check or overlap in their checks

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        self.assertEqual(len(self.pol_product_order.invoice_lines), 2, "The product line is invoiced, so it should be linked to 2 invoice lines (invoice and refund)")
+        self.assertEqual(pol_downpayment.qty_to_invoice, -1.0, "As the downpayment was invoiced separately, it will still have to be deducted from the total invoice (hence -1.0), after the refund.")
+        self.assertEqual(pol_downpayment.qty_invoiced, 1.0, "The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is")
+        self.assertEqual(len(pol_downpayment.invoice_lines), 3, "The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)")
```
if I understand this test correctly, these should be:
`"As the refund still exists, the quantity to invoice is the ordered quantity"` => `"The refund should make it so the quantity to invoice is the ordered quantity"`

`"The qty invoiced should be zero as, with the refund, the quantities cancel each other"` => `"The qty invoiced should be zero since the refund cancels the previously invoiced amount"`

`"As the downpayment was invoiced separately, it will still have to be deducted from the total invoice (hence -1.0), after the refund."` => "Downpayment was invoiced separately should still count as invoiced after refund of the product line. Since the ordered qty is 0 for down payments, this means -1 is remaining to invoice."

I would move the assert for the `qty_invoiced` for the dp lines above the `qty_to_invoice` assert since it makes the checks  easier to understand in a logical order
`""The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is""` => `"The qty invoiced should remain 1 since the refund was only for the ordered product, not the downpayment"`

"The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)" => I'm actually confused now why the downpayment is linked to the partial invoice and refund. Not sure if I reworded the ones above it correctly and I'm not sure how this one should be reworded since I don't understand if product line = downpayment line or not in this assert

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+        self.assertEqual(pol_downpayment.qty_invoiced, 1.0, "The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is")
+        self.assertEqual(len(pol_downpayment.invoice_lines), 3, "The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)")
+
+    def test_tax_and_account_breakdown(self):
```
This test could use a description. At first glance it looks like it's the same as `test_downpayment_percentage_tax_icl`

ticodoo commented on addons/purchase/tests/test_purchase_downpayment.py:
```diff
@@ -0,0 +1,686 @@
+
+        self.assertNotEqual(
+            dp_line.name, 'whatever',
+            "DP lines description should be recomputed when the linked invoice is posted",
```
`DP line's description`

ticodoo commented on addons/account/models/account_move.py:
```diff
@@ -2863,8 +2863,8 @@ def unlink(self):
                 .filtered(
                     lambda line:
-                        line.is_downpayment
-                        and line.invoice_lines <= self.mapped('line_ids')  # See https://github.com/odoo/odoo/pull/52648
```
nitpick (don't really care) unnecessary diff

ticodoo reviewed (commented):
My eyes glazed over the tax tests and I only skimmed the renaming refactorings. Seems good :+1: nice work

ticodoo commented:
@robodoo rebase-ff

Feyensv reviewed (approved):
Great work, some small comments on the fly, but nothing really critical ;)

ticodoo commented on addons/purchase/models/purchase_order_line.py:
```diff
@@ -233,10 +196,25 @@ def write(self, values):
+                line.state in ['purchase', 'done']
+                and (line.invoice_lines or not line.is_downpayment)
+                and not line.display_type
+        )
```
Hmm, it looks like you removed the exception where non-invoiced dp lines can be deleted. I didn't mean remove the condition completely, I meant add it into the existing `_unlink_except_purchase_or_done` check :thinking:

ticodoo reviewed (commented):
Ugh, Chrome doesn't seem to be handling this large PR very well. I somehow clicked on "Submit Review" by accident already, so my previous review is missing this comment:

Also just remembered a use case that will break with these new down payments. The `PurchaseRequisitionCreateAlternative._get_alternative_values()` will need to be updated to skip the down payment PO lines

R1D1CUL0US commented:
@robodoo rebase-merge

robodoo commented:
@R1D1CUL0US you can't rebase-merge. Skill issue.

ticodoo commented:
@robodoo rebase-merge r+

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

robodoo commented:
@R1D1CUL0US @ticodoo linked pull request(s) odoo/upgrade#5627 not ready. Linked PRs are not staged until all of them are ready.

d-fence commented:
robodoo priority

tivisse commented:
Hello, just discovering this PR. 

Mixin are supposed to extract transversal features that could be applied on any models, like ``mail.thread`` or ``resource.mixin``. I would rather have called it ``account.order(.line).abstract`` as the term ``mixin`` is misleading because this is only supposed to be used on 2 models.

On the other hand, we expect this kind on development to factorize the code and thus reduce the number of line of code.

R1D1CUL0US commented:
> Hello, just discovering this PR.
> 
> Mixin are supposed to extract transversal features that could be applied on any models, like `mail.thread` or `resource.mixin`. I would rather have called it `account.order(.line).abstract` as the term `mixin` is misleading because this is only supposed to be used on 2 models.
> 
> On the other hand, we expect this kind on development to factorize the code and thus reduce the number of line of code.

Hello, for your convenience, some context regarding the naming of the classes: https://github.com/odoo/odoo/pull/144608#pullrequestreview-1829400123
https://github.com/odoo/odoo/pull/144608#discussion_r1457408170
https://github.com/odoo/odoo/pull/144608#discussion_r1474463543
https://github.com/odoo/odoo/pull/144608#discussion_r1514461781
https://github.com/odoo/odoo/pull/144608#discussion_r1574678676

I personally was not aware we use 'abstract' as a keyword for naming classes in Odoo, but will keep it in mind for the future!
As for the lines of code, this is mostly due to a large number or renames of (often used) fields that were requested in this PR, as well as the adding in of a new down payment feature for purchase (= the original goal for this task) :slightly_smiling_face:

tivisse commented:
Yes, at the end, it is only naming nitpicking, but still, this could be confusing for developers.

For the number of line of codes, I didn't notice you were adding a whole new feature with a rebase-merge PR which says "introduce mixin", so don't take my comment into account.

Have a nice day !
